File: src/main/java/com/mbayou/kick4k/events/handler/KickSignatureVerifier.java
package com.mbayou.kick4k.events.handler;

import com.mbayou.kick4k.publicKey.PublicKeyClient;

import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;

public class KickSignatureVerifier {
    private final PublicKeyClient publicKeyClient;

    public KickSignatureVerifier(PublicKeyClient publicKeyClient) {
        this.publicKeyClient = publicKeyClient;
    }

    public boolean isValidSignature(String messageId, String messageTimestamp, String body, String signature) {
        try {
            String pem = this.publicKeyClient.getPublicKey();
            PublicKey key = parsePublicKey(pem);

            byte[] signedData = buildSignedData(messageId, messageTimestamp, body);
            byte[] signatureBytes = decodeBase64(signature);

            return this.verifySignature(key, signedData, signatureBytes);
        } catch (Exception exception) {
            return false;
        }
    }

    private boolean verifySignature(PublicKey publicKey, byte[] data, byte[] signatureBytes) throws Exception {
        Signature signature = Signature.getInstance("SHA256withRSA");
        signature.initVerify(publicKey);
        signature.update(data);
        return signature.verify(signatureBytes);
    }

    private static PublicKey parsePublicKey(String pem) throws InvalidKeySpecException, NoSuchAlgorithmException {
        String stripped = pem
                .replace("-----BEGIN PUBLIC KEY-----", "")
                .replace("-----END PUBLIC KEY-----", "")
                .replaceAll("\\s", "");
        byte[] keyBytes = Base64.getDecoder().decode(stripped);
        X509EncodedKeySpec spec = new X509EncodedKeySpec(keyBytes);
        KeyFactory kf = KeyFactory.getInstance("RSA");
        return kf.generatePublic(spec);
    }

    private static byte[] buildSignedData(String messageId, String timestamp, String body) {
        String data = messageId + "." + timestamp + "." + body;
        return data.getBytes(StandardCharsets.UTF_8);
    }

    private byte[] decodeBase64(String base64) {
        return Base64.getDecoder().decode(base64);
    }
}

File: src/main/kotlin/com/mbayou/kick4k/events/handler/KickSignatureVerifier.kt
package com.mbayou.kick4k.events.handler

import com.mbayou.kick4k.publicKey.PublicKeyClient
import java.nio.charset.StandardCharsets
import java.security.KeyFactory
import java.security.NoSuchAlgorithmException
import java.security.PublicKey
import java.security.Signature
import java.security.spec.InvalidKeySpecException
import java.security.spec.X509EncodedKeySpec
import java.util.Base64

class KickSignatureVerifier(private val publicKeyClient: PublicKeyClient) {
    fun isValidSignature(messageId: String, messageTimestamp: String, body: String, signature: String): Boolean {
        return try {
            val pem = publicKeyClient.getPublicKey()
            val key = parsePublicKey(pem)
            val signedData = buildSignedData(messageId, messageTimestamp, body)
            val signatureBytes = decodeBase64(signature)
            verifySignature(key, signedData, signatureBytes)
        } catch (exception: Exception) {
            false
        }
    }

    @Throws(Exception::class)
    private fun verifySignature(publicKey: PublicKey, data: ByteArray, signatureBytes: ByteArray): Boolean {
        val signature = Signature.getInstance("SHA256withRSA")
        signature.initVerify(publicKey)
        signature.update(data)
        return signature.verify(signatureBytes)
    }

    companion object {
        @Throws(InvalidKeySpecException::class, NoSuchAlgorithmException::class)
        private fun parsePublicKey(pem: String): PublicKey {
            val stripped = pem
                .replace("-----BEGIN PUBLIC KEY-----", "")
                .replace("-----END PUBLIC KEY-----", "")
                .replace("\\s".toRegex(), "")
            val keyBytes = Base64.getDecoder().decode(stripped)
            val spec = X509EncodedKeySpec(keyBytes)
            val keyFactory = KeyFactory.getInstance("RSA")
            return keyFactory.generatePublic(spec)
        }

        private fun buildSignedData(messageId: String, timestamp: String, body: String): ByteArray {
            val data = "$messageId.$timestamp.$body"
            return data.toByteArray(StandardCharsets.UTF_8)
        }

        private fun decodeBase64(base64: String): ByteArray = Base64.getDecoder().decode(base64)
    }
}
